### PR TITLE
Mintshell

### PR DESCRIPTION
**Pull request by SeHor05**

### CODE REVIEW

File: tvm_executor/src/ordinary_transaction.rs [hunk 1]
```diff
@@ -82,6 +82,8 @@ impl TransactionExecutor for OrdinaryTransactionExecutor {
         in_msg: Option<&Message>,
         account: &mut Account,
         params: ExecuteParams,
+        available_credit: i128,
```
Comment (SilkovAlexander) [inline tvm_executor/src/ordinary_transaction.rs:85]: Pls add this args to ExecuteParams

File: tvm_executor/src/transaction_executor.rs [hunk 2]
```diff
@@ -106,6 +107,8 @@ const MAX_ACTIONS: usize = 255;
 const MAX_MSG_BITS: usize = 1 << 21;
 const MAX_MSG_CELLS: usize = 1 << 13;
 
+pub type MintSet = HashMap<UInt256, u128>;
```
Comment (SilkovAlexander) [inline tvm_executor/src/transaction_executor.rs:110]: This type seems to be usused, do you need it?

File: tvm_executor/src/transaction_executor.rs [hunk 3]
```diff
@@ -816,39 +830,37 @@ pub trait TransactionExecutor {
...
+                        match acc_remaining_balance.add(&add_value) {
+                            Ok(_) => {
+                                phase.spec_actions += 1;
+                                0
+                            }
+                            Err(_) => RESULT_CODE_INVALID_BALANCE,
                         }
-                        Err(_) => RESULT_CODE_INVALID_BALANCE,
+                    } else {
+                        phase.spec_actions += 1;
```
Comment (SilkovAlexander) [inline tvm_executor/src/transaction_executor.rs:841]: MB throw error here?

File: tvm_executor/src/transaction_executor.rs [hunk 4]
```diff
@@ -816,39 +830,37 @@ pub trait TransactionExecutor {
...
                         } else {
-                            log::debug!(target: "executor", "ord values a {:?}, value {}", a, value);
-                            add_value.set_other(ECC_SHELL_KEY, value as u128)?;
+                            sub_value.set_other(ECC_SHELL_KEY, value as u128)?;
                             exchange_value = value;
                         }
                     }
-                    log::debug!(target: "executor", "exchange shell token in action in account {} with value {} and final {}", acc_remaining_balance, exchange_value, add_value);
-                    match acc_remaining_balance.sub(&add_value) {
+                    match acc_remaining_balance.grams.add(&Grams::from(exchange_value * 1_000_000))
```
Comment (SilkovAlexander) [inline tvm_executor/src/transaction_executor.rs:860]: I don't like that exchange is 1 ecc to 1e6 and mint is 1 to 1e9

File: tvm_executor/src/transaction_executor.rs [hunk 5]
```diff
@@ -862,6 +874,19 @@ pub trait TransactionExecutor {
...
                 }
+                OutAction::MintShellToken { mut value } => {
+                    if available_credit != -1 {
+                        if value as i128 > available_credit {
+                            value = available_credit.clone().try_into()?;
+                        }
+                    }
+                    acc_remaining_balance
+                        .grams
+                        .add(&(Grams::from(value) * Grams::from(1000000000)))?;
```
Comment (SilkovAlexander) [inline tvm_executor/src/transaction_executor.rs:885]: I don't like that exchange is 1 ecc to 1e6 and mint is 1 to 1e9

File: tvm_executor/src/transaction_executor.rs [hunk 6]
```diff
@@ -862,6 +874,19 @@ pub trait TransactionExecutor {
                         }
                     }
                 }
+                OutAction::MintShellToken { mut value } => {
+                    if available_credit != -1 {
```
Comment (SilkovAlexander) [inline tvm_executor/src/transaction_executor.rs:878]: `available_credit == -1 ` means infinite credit?

File: tvm_types/src/cell/slice.rs [hunk 7]
```diff
@@ -119,6 +119,10 @@ impl SliceData {
         Self { data: InternalData::None, data_window: 0..0, references_window: 0..0 }
     }
 
+    pub fn increase_start_reference(&mut self) {
```
Comment (SilkovAlexander) [inline tvm_types/src/cell/slice.rs:122]: What do you need it for? Can't you just store empty cell?

File: tvm_types/src/cell/slice.rs [hunk 8]
```diff
@@ -593,6 +597,15 @@ impl SliceData {
         Ok(value)
     }
 
+    pub fn get_next_i128(&mut self) -> Result<i128> {
+        let mut value: i128 = 0;
+        for i in 0..16 {
+            value |= (self.get_byte(8 * i)? as i128) << (8 * (15 - i));
```
Comment (SilkovAlexander) [inline tvm_types/src/cell/slice.rs:603]: why not **get_next_bits**?

File: tvm_vm/src/executor/token.rs [hunk 9]
```diff
@@ -20,7 +21,8 @@ pub(super) fn execute_ecc_mint(engine: &mut Engine) -> Status {
     engine.load_instruction(Instruction::new("MINTECC"))?;
     fetch_stack(engine, 2)?;
     let x: u32 = engine.cmd.var(0).as_integer()?.into(0..=255)?;
-    let y: VarUInteger32 = VarUInteger32::from(engine.cmd.var(1).as_integer()?.into(0..=u64::MAX)?);
+    let y: VarUInteger32 =
+        VarUInteger32::from(engine.cmd.var(1).as_integer()?.into(0..=u128::MAX)?);
```
Comment (SilkovAlexander) [inline tvm_vm/src/executor/token.rs:25]: Max value should be discussed and fixed somewhere as a const
u128::MAX * 1_000_000 is to great to fit in Grams I suppose

File: tvm_vm/src/executor/token.rs [hunk 10]
```diff
@@ -93,3 +95,12 @@ pub(super) fn execute_calculate_min_stake(engine: &mut Engine) -> Status {
     engine.cc.stack.push(int!(min_val_stake));
     Ok(())
 }
+
+pub(super) fn execute_mint_shell(engine: &mut Engine) -> Status {
+    engine.load_instruction(Instruction::new("MINTSHELL"))?;
+    fetch_stack(engine, 1)?;
+    let x: u64 = engine.cmd.var(0).as_integer()?.into(0..=2147483647)?;
```
Comment (SilkovAlexander) [inline tvm_vm/src/executor/token.rs:102]: Max value should be discussed and fixed somewhere as a const

File: tvm_executor/src/ordinary_transaction.rs [hunk 11]
```diff
@@ -82,6 +82,8 @@ impl TransactionExecutor for OrdinaryTransactionExecutor {
...
+        minted_shell: &mut u128,
```
Comment (SilkovAlexander) [inline tvm_executor/src/ordinary_transaction.rs:86]: This parameter may be changed to output

**Review by SilkovAlexander — APPROVED** [review]

**Review by elasticLove1 — APPROVED** [review]

**Pull request merged**